Address Codegen bug related to marking subregister MachineOperand defines as undef

bababuck opened the pull request:
Fixes #130884.

In the failing case, the following COPIES are moved such that they swap in order:
```
  96B INLINEASM &"; def $0, $1" [attdialect], $0:[regdef], implicit-def $vgpr4, $1:[regdef], implicit-def $vgpr5
  112B undef %18.sub1:vreg_64 = COPY killed $vgpr4
  128B %18.sub0:vreg_64 = COPY killed $vgpr5
```
However, the `undef` flag is improperly left of the 112B instrcution, resulting in:
```
  96B INLINEASM &"; def $0, $1" [attdialect], $0:[regdef], implicit-def $vgpr4, $1:[regdef], implicit-def $vgpr5
  128B %18.sub0:vreg_64 = COPY killed $vgpr5
  112B undef %18.sub1:vreg_64 = COPY killed $vgpr4
```
This causes the liveness verifier to flag this as an error.

The proper result code should be:
```
  96B INLINEASM &"; def $0, $1" [attdialect], $0:[regdef], implicit-def $vgpr4, $1:[regdef], implicit-def $vgpr5
  128B undef %18.sub0:vreg_64 = COPY killed $vgpr5
  112B %18.sub1:vreg_64 = COPY killed $vgpr4
```
 
Notes:
- An alternative solution would be to prevent the re-ordering in this situation.

- I made the change inside `reschedulePhysReg()`, but we could instead move the change up the call tree to `moveInstruction()`, thoughts? I avoided doing that to limit the scope of this change.

- Passes lit and unit tests.

@arsenm

bababuck committed:
Revert "AMDGPU: Disable machine verifier in failing test"

This reverts commit bf2d1c46072a0461cb3ddcaefcafeccc2637995e.

bababuck commented:
Update: Fixed clang-format errors.

bababuck commented:
The format check is flagging the use of `undef` in some of my comments.

Is this MR is okay with respect to the deprecation of `undef`?

arsenm commented on llvm/lib/CodeGen/MachineScheduler.cpp:
```diff
@@ -3979,6 +3979,32 @@ void GenericScheduler::reschedulePhysReg(SUnit *SU, bool isTop) {
+      }
+    }
+    if (SubregDefs.size()) {
+      for (auto CurrInst = InsertPos; CurrInst != Copy; ++CurrInst) {
```
You shouldn't need to loop through all the instructions here. You should rely on the LiveIntervals to check if the lanes are undefined or not, after they've been updated in moveInstruction (although actually, maybe LiveIntervals::handleMove ought to be taking care of this in the first place)

bababuck replied:
I moved the new functionality into LiveIntervals::handleMove(), the code now uses the LiveRanges to find the instruction to compare with. I pushed this new version but I'm not sure it's fully correct.

The code looks for the case like the following (pseudo-IR):
```
undef %0.sub0 = SOME_FUNCTION %1
...
%0.sub1 = SOME_FUNCTION %2
```
gets changes to
```
%0.sub1 = SOME_FUNCTION %2
undef %0.sub0 = SOME_FUNCTION %1
```
and it changes it to 
```
undef %0.sub1 = SOME_FUNCTION %2
undef %0.sub0 = SOME_FUNCTION %1
```

However, it doesn't handle a case like the following:
```
%0.sub0 = SOME_FUNCTION %1
...
undef %0.sub1 = SOME_FUNCTION %2
%0.sub2 = SOME_FUNCTION %3
```
getting changed too
```
%0.sub2 = SOME_FUNCTION %3
%0.sub0 = SOME_FUNCTION %1
...
undef %0.sub1 = SOME_FUNCTION %2
```
The reasoning is that the starting IR of that case is already erroneous, but I'm not confident in that analysis.

An alternative would be to look at every LiveInterval::Segment for an `undef` between the old and new location of the instruction.

In addition, this only handled the `Up` case, not the case where an `undef` instruction is moved `Down` in the program to after a non-`undef` definition of a sub-register.

arsenm replied:
The starting MIR is correct, the new output with 2 undefs is wrong. When the first undef subregister def moves past the second, the undef flag needs to move to the first subregister def. They both should not end up with the undef flag

arsenm commented on llvm/lib/CodeGen/MachineScheduler.cpp:
```diff
@@ -3979,6 +3979,32 @@ void GenericScheduler::reschedulePhysReg(SUnit *SU, bool isTop) {
+    // Check to make sure that there are no subreg defintions of the given
+    // register between it's new and old location that are marked as undef. If
+    // so, mark the current instruction as undef instead.
+    SmallVector<MachineOperand *, 1> SubregDefs;
```
Shouldn't need to accumulate this set

arsenm commented on llvm/lib/CodeGen/MachineScheduler.cpp:
```diff
@@ -3979,6 +3979,32 @@ void GenericScheduler::reschedulePhysReg(SUnit *SU, bool isTop) {
+          if (MO.isReg() && MO.isDef() && MO.isUndef() && MO.getSubReg() != 0) {
+            for (auto *MISubregDef : SubregDefs) {
+              if (MISubregDef->getReg() == MO.getReg()) {
+                assert(!MISubregDef->isUndef() &&
```
There might be a hazard with overlapping subregister defs

bababuck replied:
I've seen in some of the tests (notably the test that changed with my latest change) situations like this where:
```
undef %0.sub0 = SOME_FUNCTION %1
undef %0.sub1 = SOME_FUNCTION %2
```
Why is this considered legal/desired. My understanding is that without the `undef` the other sub-registers are considered un-modified. However, with the `undef` the other sub-registers are now undefined, so after this sequence `%0.sub0` would be undefined. Wouldn't that violate SSA constraints since `%0.sub0` is assigned twice, once with a value and secondly as `undef`.

arsenm replied:
Subregister defs only exist post-SSA, where the scheduler runs

arsenm reviewed (commented):
Should get a dedicated MIR test

arsenm commented:
> Is this MR is okay with respect to the deprecation of `undef`?

Yes

bababuck committed:
Add MIR test targeting #130884

bababuck committed:
Properly handle undef statements in ScheduleDagMI::moveInstruction

If a subregister def is moved such that it is not live prior to
that instruction, mark as Undef. Also remove any Undef flag from the next
def of this register.

Fixes #130884.

arsenm commented on llvm/lib/CodeGen/LiveIntervals.cpp:
```diff
@@ -1570,6 +1570,29 @@ void LiveIntervals::handleMove(MachineInstr &MI, bool UpdateFlags) {
 
   HMEditor HME(*this, *MRI, *TRI, OldIndex, NewIndex, UpdateFlags);
   HME.updateAllRanges(&MI);
+
```
I'm not 100% sure if handleMove should be responsible for updating the actual MI. updateAllRanges conservatively strips kill flags (which is kind of annoying), but seems to not touch the MIR

bababuck replied:
Would moving it back to `ScheduleDagMI::moveInstruction` be better in your opionion? We have access to a `LiveIntervals` object there, so the same logic could be used.

arsenm commented on llvm/lib/CodeGen/LiveIntervals.cpp:
```diff
@@ -1570,6 +1570,29 @@ void LiveIntervals::handleMove(MachineInstr &MI, bool UpdateFlags) {
+  // Check to make sure that there are no subreg defintions marked undef
+  // after the moved operation. If so, mark the current instruction as undef
+  // instead.
+  for (MachineOperand &MO : MI.operands()) {
```
In theory the LiveIntervals are correct at this point, and the MIR is out of sync. It would probably be easier to express this in terms of operand-is-not-undef-but-not-live-at-this-point

bababuck replied:
I did do some more digging and was able to better utilize the `MachineInst` functions to clean up some of the code. 

I think that expressing this in terms of `operand-is-not-undef-but-not-live-at-this-point` does make sense because that is more similar to how the verifier works.
